ride_share.rb

```diff
@@ -0,0 +1,103 @@
+# Step 3: Make the data structure!
+
+# Setup the entire data structure:
+# based off of the notes you have above, create the
+# and manually write in data presented in rides.csv
+# You should be copying and pasting the literal data
+# into this data structure, such as "DR0004"
+# and "3rd Feb 2016" and "RD0022"
+
+# the dates needed a bit of tweeking to be apropriately ordinalized.
+
+# based on layering (nesting) method 3:
+all_rides = {
+  :DR0001 => [
+    {:DATE => "3rd Feb 2016", :COST => 10, :RIDER_ID => "RD0003", :RATING => 3},
+    {:DATE => "3rd Feb 2016", :COST => 30, :RIDER_ID => "RD0015", :RATING => 4},
+    {:DATE => "5th Feb 2016", :COST => 45, :RIDER_ID => "RD0003", :RATING => 2}
+  ],
+  :DR0002 => [
+    {:DATE => "3rd Feb 2016", :COST => 25, :RIDER_ID => "RD0073", :RATING => 5},
+    {:DATE => "4th Feb 2016", :COST => 15, :RIDER_ID => "RD0013", :RATING => 1},
+    {:DATE => "5th Feb 2016", :COST => 35, :RIDER_ID => "RD0066", :RATING => 3}
+  ],
+  :DR0003 => [
+    {:DATE => "4th Feb 2016", :COST =>  5, :RIDER_ID => "RD0066", :RATING => 5},
+    {:DATE => "5th Feb 2016", :COST => 50, :RIDER_ID => "RD0003", :RATING => 2}
+  ],
+  :DR0004 => [
+    {:DATE => "3rd Feb 2016", :COST =>  5, :RIDER_ID => "RD0022", :RATING => 5},
+    {:DATE => "4th Feb 2016", :COST => 10, :RIDER_ID => "RD0022", :RATING => 4},
+    {:DATE => "5th Feb 2016", :COST => 20, :RIDER_ID => "RD0073", :RATING => 5}
+  ]
+}
+
+########################################################
+# Step 4: Total Driver's Earnings and Number of Rides
+
+rides_num = {} 
+all_rides.each do |driver, rides|
+  rides_num[driver] = rides.length
+end
+
+driver_total_income = {}
+all_rides.each do |driver, rides|
+  driver_income = rides.map do |each_ride|
+    each_ride [:COST]
+  end
+  driver_total_income[driver] = driver_income.sum
+end
+
+driver_average_rating = {}
+all_rides.each do |driver, rides|
+  driver_rating = rides.map do |each_ride|
+    each_ride[:RATING].to_f
+  end
+  driver_average_rating[driver] = ( driver_rating.sum / rides.length ).round(2)
+end
```

Comment:
You iterate over `all_rides` a lot, with a different goal each time. Is there an opportunity to DRY up this code with a single loop of `all_rides` and maybe a few methods?